### PR TITLE
text_generation_utils memory reduction if no logprob needed

### PR DESCRIPTION
**Pull request by yzhang123**
# What does this PR do ?

make text_generation_utils.py more memory efficient for inference especially for long context sequences 
* by making logprob computation optional

**Collection**: nlp

# Changelog 
- change peft eval model class for restoring to MegatronGPTSFTModel, since NLPModel does not have _build_model
- removing logprob computation in text_generation_utils since this is very memory consuming because output embedding are all gathered for this. reusing parameter compute_logprob to check if this is needed, by default false

# Usage

```python
 python examples/nlp/language_modeling/megatron_gpt_eval.py \
    ++trainer.devices=${DEVICES} \
    gpt_model_file=$CKPT \
    trainer.num_nodes=1 \
    inference.greedy=True \
    pipeline_model_parallel_size=$PP \
    tensor_model_parallel_size=$TP \
    trainer.precision=bf16 \
    inference.compute_logprob=True \

python examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py \
 inference.compute_logprob=True ..
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/text_generation_strategy.py [hunk 1]
```diff
@@ -67,7 +66,7 @@ def forward_step(self, batch, tensor_shape):
 
         return output_tensor
 
-    def tokenize_batch(self, sentences, max_len, add_BOS):
+    def tokenize_batch(self, sentences, max_len, add_BOS, truncate_prompt_length):
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/text_generation_strategy.py:69]: Can you add this to the docstring?

File: nemo/collections/nlp/modules/common/text_generation_utils.py [hunk 2]
```diff
@@ -317,18 +323,19 @@ def receive_generate_info():
     """
     model_parallel_group = parallel_state.get_model_parallel_group()
     src = get_model_parallel_src_rank()
-    input_info_tensor = torch.empty(10, dtype=torch.float32, device=torch.cuda.current_device())
+    input_info_tensor = torch.empty(11, dtype=torch.float32, device=torch.cuda.current_device())
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/text_generation_utils.py:326]: added compute_logprob as new entry to input_info_tensor, hence need to increase by one. added comment what compute_logprob does

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 3]
```diff
@@ -238,9 +238,11 @@ def main(cfg) -> None:
     except AttributeError:
         pass
 
+    assert cfg.inference.truncate_prompt_length == -1 or cfg.inference.truncate_prompt_length >= 0
```
Comment (arendu) [inline examples/nlp/language_modeling/megatron_gpt_eval.py:241]: so just >= -1?

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 4]
```diff
@@ -265,9 +267,10 @@ def main(cfg) -> None:
 
     # Second method of running text generation, call trainer.predict
     ds = RequestDataSet(OmegaConf.to_container(cfg.prompts))
-    request_dl = DataLoader(dataset=ds, batch_size=2)
+    request_dl = DataLoader(dataset=ds, batch_size=1)
```
Comment (yidong72) [inline examples/nlp/language_modeling/megatron_gpt_eval.py:270]: is this change necessary?

**Review by yidong72 — COMMENTED** [review]
looks good overall. 
why we need truncate_prompt_length argument?

**Comment by yzhang123** [comment]
> looks good overall. why we need truncate_prompt_length argument?

if anyone want's to truncate their input to a maximum length? but i can see why this not a necessary change, your call if you want me to remove it. I'm ok either way

File: examples/nlp/language_modeling/megatron_gpt_eval.py (hunk 4 above)
Comment (yzhang123) [reply examples/nlp/language_modeling/megatron_gpt_eval.py:270]: > looks good overall. why we need truncate_prompt_length argument?

If someone want's to truncate their prompt, like with p-tuning truncating the context. but this is not a deal breaker, so your call if you want to remove it

Comment (yzhang123) [reply examples/nlp/language_modeling/megatron_gpt_eval.py:270]: ok i can revert the bs

**Review by arendu — CHANGES_REQUESTED** [review]
log prob changes looks great! just suggested @yzhang123  to not include the truncation logic in this PR.

**Comment by yzhang123** [comment]
> log prob changes looks great! just suggested @yzhang123 to not include the truncation logic in this PR.

this has been removed from the pr

**Review by arendu — APPROVED** [review]
LGTM!
